### PR TITLE
[SPARK-32968][SQL] Prune unnecessary columns from CsvToStructs

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to do column pruning for CsvToStructs expression if we only require some fields from it.

### Why are the changes needed?

`CsvToStructs` takes a schema parameter used to tell CSV Parser what fields are needed to parse. If `CsvToStructs` is followed by GetStructField. We can prune the schema to only parse certain field.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by viirya** [comment]
cc @HyukjinKwon @dongjoon-hyun @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:38]: Thank you for generalizing this. This renaming looks fine to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:58]: indentation?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:112]: Just a question. Do you have more optimizations in this area in the near future?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:108]: Although this comes from the old file, shall we remove this empty line at this PR?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (except two minor style comments).

Thanks, @viirya .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:112]: Yea, I'll add a few more Csv optimization similar to Json ones.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:52]: What about:

```scala
case p if SQLConf.get.jsonExpressionOptimization => p.transformExpressions(jsonOptimization)
case p if SQLConf.get.csvExpressionOptimization => p.transformExpressions(csvOptimization)
case p => p
```
? Or is the optimization dependent on each other?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:55]: nit but maybe private.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:52]: Not dependent, but if so `p` only goes to match first case when both are enabled.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:61]: @viirya, I know this is not the part of this PR but nit: `v.nameExprs.head.semanticEquals(c.valExprs.head.children.head)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:52]: Ah I see. Since we generalized the rule, it might more make sense to have one configuration to enable and disable this ..

**Comment by HyukjinKwon** [comment]
Should we maybe separate the refactoring out of this PR? Now we have one generalized rule but have two configurations to turn on and off each.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:110]: here too. nit but maybe private

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:52]: To enable/disable entire rule, `spark.sql.optimizer.excludedRules` should be able to help. So the purpose here is to have individual configs for enabling csv/json.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:114]: @viirya, this has a behaviour about parse mode. For example, previously it worked in failfast mode when other columns can be passed but. Looks like now it fails, for example:

```scala
spark.sparkContext.parallelize(Seq("a,b")).toDF("csv")
  .selectExpr("from_csv(csv, 'a string, b string', map('mode', 'failfast')) as parsed")
  .selectExpr("parsed.a").show()
```

Before this PR:

```
+---+
|  a|
+---+
|  a|
+---+
```

After this PR:

```
org.apache.spark.SparkException: Malformed records are detected in record parsing. Parse Mode: FAILFAST. To process malformed records as null result, try setting the option 'mode' as 'PERMISSIVE'.
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvExprsSuite.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvExprsSuite.scala:29]: I think we should have an e2e test too since there's runtime options available in CSV that affects output and how to parse (e.g., https://github.com/apache/spark/pull/30912/files/4353f4efa054980de0a2905dbcf7000fd32524e4#r548365520).

Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvExprsSuite.scala:29]: Okay. I will add e2e test later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:52]: Hm .. okay. I still prefer to separate refactoring from this PR but it's fine if both you or @dongjoon-hyun prefer.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala]: Good catch! `from_csv` needs to pass in both data schema and required schema for pruning usage.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala]: Hm, rethinking about it, I think it is safer to do pruning only if `options` is empty. This optimization might hide a parsing error when parsing entire schema under "failfast" mode.

Although it is arguable if it still should throw the same error if we only need a field that could be parsed correctly. But I don't want to change the behavior before the optimization.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvExprsSuite.scala
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvExprsSuite.scala:29]: Added e2e tests now. Thanks.

**Comment by viirya** [comment]
For Json's e2e tests, I will work in other PR. Thanks.

**Comment by viirya** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala]: I re-checked Json optimization. Json parser behavior seems different than Csv here. Pruned schema doesn't change failfast behavor of json parser.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:52]: Technically, I'm neutral in terms of separating.

**Review by dongjoon-hyun — APPROVED** [review]
It seems there is a discussion on the refactoring, I leave this to you guys, @viirya and @HyukjinKwon . I'm okay for both ways.

**Comment by HyukjinKwon** [comment]
I'm okay but allow me to review one more time before merging it in. CSV/JSON ones have many corner cases which are easy to overlook ..

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:113]:
```suggestion
    case g @ GetStructField(c @ CsvToStructs(schema: StructType, _, _, _, None), ordinal, _)
        if schema.length > 1 && c.options.isEmpty =>
      val prunedSchema = StructType(Seq(schema(ordinal)))
      g.copy(child = c.copy(requiredSchema = Some(prunedSchema)), ordinal = 0)
```

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:113]: @viirya, shall we exclude this optiomization when malformed column is selected?

```scala
spark.sparkContext.parallelize(Seq("a,b,c,d")).toDF("csv")
  .selectExpr("from_csv(csv, 'a string, b string, _corrupt_record string') as parsed")
  .selectExpr("parsed._corrupt_record").show()
```

Before:

```
+---------------+
|_corrupt_record|
+---------------+
|        a,b,c,d|
+---------------+
```

After:
```
java.lang.ArrayIndexOutOfBoundsException: 0
	at org.apache.spark.sql.catalyst.expressions.GenericInternalRow.genericGet(rows.scala:201)
	at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.getAs(rows.scala:35)
	at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.get(rows.scala:37)
	at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.get$(rows.scala:37)
	at org.apache.spark.sql.catalyst.expressions.GenericInternalRow.get(rows.scala:195)
	at org.apache.spark.sql.catalyst.util.FailureSafeParser.$anonfun$toResultRow$2(FailureSafeParser.scala:47)
	at scala.Option.map(Option.scala:230)
```

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala]: Hm, interesting. I will also add test for this.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala]: Just added a test. I also checked json. Json parser seems smarter and doesn't have the same issue.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:39]: no big deal but I would do this as a `private lazy val` in case it doesn't respect runtime config.

**Comment by HyukjinKwon** [comment]
I have no more comments for now. I don't mind if somebody merges FWIW.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala:113]:
```suggestion
        if schema.length > 1 && c.options.isEmpty && schema(ordinal).name != nameOfCorruptRecord =>
        // When the parse mode is permissive, and corrupt column is not selected, we can prune here.
        // from `GetStructField`. To be more conservative, it does not optimize when any option
        // is set.
```

**Review by HyukjinKwon — COMMENTED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Last commit was only comment. I manually ran linter locally and verified that it passed. I am going to merge this.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by viirya** [comment]
Thanks @dongjoon-hyun @HyukjinKwon

**Pull request closed without merging**
